src/plugins/grpc/client.js

```diff
@@ -0,0 +1,100 @@
+'use strict'
+
+const { FORMAT_HTTP_HEADERS, Tags } = require('opentracing')
+
+function patch (grpc, tracer, config) {
+  if (config.client === false) return
+
+  config = config.client || config
+
+  function gRPCInterceptor (options, nextCall) {
+    const { path, responseStream } = options.method_definition
+    const scope = tracer.scopeManager().active()
+    const span = tracer.startSpan('grpc.client', {
+      childOf: scope && scope.span(),
+      tags: {
+        [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_CLIENT,
+        'grpc.method_name': path,
+        'grpc.method_type': responseStream ? 'server_stream' : 'unary',
+        'resource.name': path.split('/').slice(-1),
+        'service.name': config.service || `${tracer._service}-grpc`
+      }
+    })
+
+    return new grpc.InterceptingCall(nextCall(options), {
+      sendMessage: (message, next) => {
+        if (config.arguments) {
+          const values = config.arguments(message)
+
+          for (const key in values) {
+            span.setTag(`grpc.arguments.${key}`, values[key])
+          }
+        } else {
+          span.setTag(`grpc.arguments`, message)
+        }
+
+        next(message)
+      },
+
+      start: (metadata, _listener, next) => {
+        if (metadata) {
+          for (const [key, value] of Object.entries(metadata.getMap())) {
+            span.setTag(`grpc.metadata.${key}`, value)
+          }
+        }
+
+        // Inject tracing headers into the grpc call's metadata.
+        const tracingMetadata = metadata || new grpc.Metadata()
+        const meta = {}
+
+        tracer.inject(span, FORMAT_HTTP_HEADERS, meta)
+
+        for (const [key, value] of Object.entries(meta)) {
+          tracingMetadata.set(key, value)
+        }
+
+        next(tracingMetadata, {
+          onReceiveStatus: (status, next) => {
+            if (status.code !== 0) {
+              span.addTags({
+                'error.code': status.code,
+                'error.msg': status.details
```

Comment:
> This will conflict with the predefined error object that we usually send. Any reason not to use status.code and status.details directly, and simply set error to true?

🤔 I was attempting to provide the functionality of displaying the correct error message in Datadog by setting `error.code` and `error.msg` here. Is that not how it works? (ex: from the [graphql plugin](https://github.com/DataDog/dd-trace-js/blob/master/src/plugins/graphql.js#L316-L322))

> These could probably always be included as well if they don't contain the word "error".

I could include them but `status.code` is the gRPC code which will always be `0` for a successful message (see [here](https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/constants.js#L32) and [here](https://github.com/grpc/grpc-node/blob/633736225ac4f8ce05bd246ffe576616e9290615/packages/grpc-native-core/src/client_interceptors.js#L84)) so I didn't really find it interesting to include. Happy to add it if you feel it's useful.

Also, `status.details` is only set if there's a non-zero status code.

Reply:
I agree that `status.code` and `status.details` are not necessary in that context. We can always add them later if necessary.

About the error tags, basically our backend extracts `error.type`, `error.type` and `error.stack` as metadata and removes the tags. This means that the result of the above will be an `error.code` tag and an error message in the UI. I'm not 100% sure if `error.type` is required, but we can always add it later if the error doesn't appear as expected.

Reply:
Actually I would probably go with `status.code` + `error.msg` in that case, otherwise it seems like there was some kind of error object. `status.code` seems to be more semantically correct. It's actually a status code, the fact that it represents an error is irrelevant for that tag specifically. Basically the same logic as `http.status_code`.

Reply:
Makes sense. I'll update.